Move connectee_name property out of AbstractConnector; Serialize Input connectee_name

chrisdembia opened the pull request:
### Background

This PR moves the `connectee_name` property from `AbstractConnector` to `Component`, and also now causes Inputs' connectee_name to be serialized and deserialized.

The work in this PR originates from PR #1012, whose aim was to serialize Inputs' connectee_name in the same manner as Connectors' connectee_name. In doing so, however, I ran into an issue with that serialization scheme. As a result, this PR changes the way in which the connectee_name is serialized, and also provides some additional benefits (see #359, #825, #1009).
### Change in XML connectee name syntax

Here's the previous syntax:

``` xml
<Geometry>
   <connectors>
      <Connector_PhysicalFrame_ name="frame">
          <!--Name of the component this Connector should be connected to.-->
          <connectee_name>ground</connectee_name>
      </Connector_PhysicalFrame_>
  </connectors>
</Geometry>
```

Here's the new syntax:

``` xml
<MyComponent>
  <!--Path to a Component to satisfy the Connector 'childFoo' of type Foo (description: The frame to which this geometry is attached. Note, being connected to a Frame means its transform is used to position this Geometry.).--> 
  <connector_frame_connectee_name>ground</connector_frame_connecteee_name>
</MyComponent>
```
#### Advantages in the new syntax and implementation:
1. Shorter (fewer lines); especially if the comment is removed.
2. The user input is limited to exactly what must be provided (user cannot put in the wrong type (`PhysicalFrame`), or wrong connector name (`frame`)).
3. The XML comment is specific to the connector (rather than generic), and includes the comment provided in `OpenSim_DECLARE_CONNECTOR`.
4. Removes the need tor register (`Object::registerType`) a Connector/Input for each new template argument.
5. The user could previously reorder the Connectors within the `connectors` property list, or add non-existant connectors (e.g., could add a `Connector_Foo_` and it would get deserialized, even though Foo shouldn't have such a connector).
6. I think that users _should_ be able to directly edit XML files (though I know others disagree), and this new syntax is hopefully easier to view/edit.
#### Disadvantages of the new syntax and implementation:
1. The new implementation relies on more macros, which are confusing to users (plugin developers).
2. Slight abuse of the property system (introducing two new macros to generate properties that are specific to connectee names.
3. In XML, the inputs and connectors are no longer grouped together anymore. Previously, all connectors and inputs showed up together at the top of a Component's xml element. Also, we could update the serialization code to serialize connectors before all other properties.
4. If we learn, down the road, that we actually want more properties associated with a connector other than just its connectee_name, then we could end up in a somewhat messy situation. But this is a hypothetical concern.
### What does doxygen look like?

Here's the top of the doxygen page for Geometry, which has a list input and a connector:

![image](https://cloud.githubusercontent.com/assets/846001/19335553/081af4fc-90b9-11e6-997d-992174cc3cf0.png)

The `input_transform_connectee_name` property does not show up, as expected. However, the detailed description of the `transform` Input describes the `<input_transform_connectee_name>` XML element:
![image](https://cloud.githubusercontent.com/assets/846001/19335515/c3e1709a-90b8-11e6-9a73-ca7ec01e0fce.png)

Here's an example detailed description for a connector:
![image](https://cloud.githubusercontent.com/assets/846001/19335537/ebc33422-90b8-11e6-83d0-a3795a8fa369.png)
### Summary of changes to the code

Note: Unfortunately, I was not very good about staying focused to the task at hand, and this PR includes a bunch of other smaller changes.
1. Wrote new property-generating macros, `OpenSim_DECLARE_PROPERTY_CONNECTEE_NAME` and `OpenSim_DECLARE_LIST_PROPERTY_CONNECTEE_NAME`. These generate `std::string` properties, but do _not_ generate a bunch of extra functions like `get_prop_name()`. As a result, the only way to edit the connectee name is through the Input. The C++ interface has not really changed (caveat: advanced users can still edit _any_ property via the property table).
   1. This property is constructed using an in-class member initializer, and thus it is not required to call `constructProperty_prop_name()` in the constructor.
2. Gave `AbstractConnector` access to the new property generated in the macro.
3. `Component::getInput()` and `Component::getConnector()` now have a "side effect" of setting the "owner" on the input/connector before returning it. This is to deal with the fact that the input's/connector's "owner" pointer is lost after copying the component. This is less than idea but is way better than writing a custom copy constructor for Component.
4. Wrote XML conversion code by implementing `updateFromXMLNode()` for `Component`, to update previous XML files to the new format. I bumped up the `XMLDocument::LatestVersion` to 30508.
5. Created more documentation within the `CONNECTOR` and `INPUT` macros.
6. Wrote tests to make sure that serializing -> deserializing input connectee names works.
7. Improved testComponents so that it now wires up inputs.

Unrelated changes:
1. Rolled back the model file `Applications/Analyze/test/subject01_adjusted_spring.osim` from version 30503 to 30000. Here's part of the commit message for that commit:
   
   > This commit reverts the model used in the testStaticOptimization passive force
   > test to version 30000. Previously, the model was at version 30503, but we
   > changed the Station connector's name from reference_frame to parent_frame after
   > 30503 was first created.
2. Made constructors for Connectors/Inputs protected, so that only Components will construct them.
3. Rearranged the implementation of `Input::connect` methods. This causes a really messy diff, but there are no significant changes here. I mostly just reorganized the code so I could reduce indentation.
### Looking forward

I am interested in further improving the XML syntax, as described in #1009 (thanks @tkuchida). I think we _could_ achieve the following:

``` xml
<MyComponent>
  <input_desired_angle_connectee_name>
    <InputConnecteeName path="../marker_data" output="column" alias="right_ankle"/>
  </input_desired_angle_connectee_name>
</MyComponent>
```

I don't want to tackle this here though.

chrisdembia commented:
I have finished updating the PR description, and this is now ready for review. 

(Note to @tkuchida: this PR likely conflicts with your #1316. I want your PR to be merged first, so that I can be the one to fix the conflicts)

chrisdembia commented:
You can view the doxygen for this PR on the myosin website: http://myosin.sourceforge.net/1340/classOpenSim_1_1Geometry.html

chrisdembia commented on OpenSim/Common/ComponentConnector.h:
```diff
@@ -200,45 +175,110 @@ class OSIMCOMMON_API AbstractConnector : public Object {
+    print out a warning. */
+    void finalizeFromProperties() {
+        // TODO This check may go elsewhere once the connectee_name
+        // property is a ComponentPath (or a ChannelPath?).
```
Fortunately, `:`, `(`, and `)` are not forbidden characters in `ComponentPath`, so yeah I can try that.

chrisdembia commented on OpenSim/Common/ComponentConnector.h:
```diff
@@ -200,45 +175,110 @@ class OSIMCOMMON_API AbstractConnector : public Object {
+    
+    /** Check if entries of the connectee_name property contain spaces; if so,
+    print out a warning. */
+    void finalizeFromProperties() {
```
How about a more specific name like `checkConnecteeNameProperty()`? Keep in mind that this function should disappear if we make the property type into InputConnecteePath or ChannelPath instead of std::string.

aseth1 replied:
I like the more specific name. Can you also place a TODO to remove the method once we move to storing Paths instead of strings?

chrisdembia replied:
Done.

chrisdembia commented on OpenSim/Common/Component.h:
```diff
@@ -2180,18 +2184,21 @@ OpenSim_DECLARE_ABSTRACT_OBJECT(Component, Object);
-        _connectorsTable[name] = ix;
-        return ix;
+    bool constructConnector(const std::string& name,
+                            const PropertyIndex& connecteeNameIndex) {
```
I started to attempt this and I think it will work! Good idea. The biggest question is whether doing so would preclude creating separate input/connector macros in the case that the developer wants a custom copy constructor. In such a case, the developer will need to call macro-generated methods like `constructInput_input_name()` in their constructors. I think it should work fine, but I just haven't tried it yet.

aseth1 replied:
Great! I think calling macro generated methods in custom copy would be fine and we already have a precedent for that with `copyProperty_property_name` so I think it would at least be consistent.

chrisdembia commented on OpenSim/Common/Component.h:
```diff
@@ -813,7 +811,19 @@ OpenSim_DECLARE_ABSTRACT_OBJECT(Component, Object);
+                // The `this` pointer must be non-const because the Connector
+                // will want to be able to modify the connectee_name property.
+                const_cast<AbstractConnector*>(it->second.get())->setOwner(
+                        const_cast<Self&>(*this));
```
I agree, and that was my first plan. But then I ran into the nice code from the README:

```cpp
Ellipsoid bodyGeometry(0.1, 0.5, 0.1);
...
humerusCenter->attachGeometry(bodyGeometry.clone()); <--- no chance to call finalizeFromProperties()
```

This causes the following to be invoked:

```cpp
bodyGeometry.clone()->updConnector<Frame>("frame").setConnecteeName(...);
```

But since we're working with a copy of the original `bodyGeometry`, calling `setConnecteeName()` doesn't work as the copy doesn't know about its parent yet.


These are the options I've thought of so far:
1. Prohibit the concise syntax above and require calling `finalizeFromProperties()`.
2. Cause the same behavior for subcomponents; e.g., have `getComponent()`, `getComponentList()`, etc invoke `setParent()` before returning the subcomponent.
3. When a user calls `Geometry::setFrame()`, or more generally sets up connections via the API, do not store the connectee name at that time; hold onto the given reference. Only fill in the connectee_name property when serializing, etc. This change would have other benefits too (avoid issues with connecting to an object that isn't part of the model yet). However, this is also out of scope of this PR.

Another possibility to

aseth1 replied:
I see now, thanks. My preference then is 2. if it works consistently. I also think 3. would be great and addresses the ability to rename components. If the reference is the connection then we could extract the name for the purpose of serialization. 

For now let's leave your solution for Connectors/Inputs. After the PR is merged lets try 2. to establish consistency with components. Let's put 3. on the back-burner, for now.

chrisdembia replied:
> After the PR is merged lets try 2. to establish consistency with components. 

I've opened #1393 to track this.

aymanhab replied:
Should probably add this to #1393 but having  a side-effect inside a getter method (that users would assume is const/safe) is less desirable, so I'd rather see option 3 but that's futuristic for now.

chrisdembia replied:
@aymanhab I have added your comment to #1393. For this PR, would it be sufficient to add a note to the function doxygen block saying that there is such a side effect?

chrisdembia commented on OpenSim/Common/Component.h:
```diff
@@ -706,34 +714,24 @@ OpenSim_DECLARE_ABSTRACT_OBJECT(Component, Object);
-            names[i] = get_connectors(i).getName(); 
+        std::vector<std::string> names;
+        for (const auto& it : _connectorsTable) {
+            names.push_back(it.first);
```
> now it is the order maintained by the map

Correct.

> so this might be OK.

I haven't thought of a problem with this yet. But if there is, we can change the data structure from a map to a vector.

aymanhab replied:
Thanks for clarifying the order issue here @chrisdembia , you can ignore my earlier comment except for uniqueness guarantee.

chrisdembia commented on OpenSim/Common/Component.cpp:
```diff
@@ -1035,6 +1033,25 @@ void Component::setNextSubcomponentInSystem(const Component& sub) const
+            //           </connectors>
+            // New:      <connector_parent_connectee_name>...
+            //               </connector_parent_connectee_name>
+            //
```
> which isn't totally clear about what is auto generated and what is user defined.

The comment is completely auto-generated. When the user gets a default XML representation for a component from the GUI's XML Browser (or from `opensim-cmd print-xml`), they will see:

```xml
<MyComponent>
  <!--Path to a Component to satisfy the Connector 'childFoo' of type Foo (description: The frame to which this geometry is attached. Note, being connected to a Frame means its transform is used to position this Geometry.).--> 
  <connector_frame_connectee_name></connector_frame_connecteee_name>
</MyComponent>
```

So the only thing that is user-defined is the value of the `<connector_frame_connectee_name>` element.


> but if the comments disappear (e.g. MATLAB XML editing) then determining the type that satisfies "parent" can be challenging.

It may be possible to make the field name contain the type: `<connector_PhysicalFrame_frame_connectee_name>`. I'm not sure if this is possible to implement, and might look bad for templated types (e.g., `<connector_TableSource_double__frame_connectee_name>`.

While I understand the sentiment of more directly putting the type name in the XML tags, I'm not sure it's the right thing to do. The type of the Connector is indeed documentation, not user-editable. So I think it is a benefit that the type name does not appear "user-editable" in the XML file. For example, users might mis-specify a Connector to Body with the erroneous tag `<Connector_Ground_ name="frame">`. 

I think it is reasonable to expect that the comments are necessary for users to know how to fill in the XML fields. I think we already hold this expectation for our XML files. For most existing properties, I would not feel comfortable editing them without seeing the XML comment (i.e., Body's `<inertia>`, Muscle's `<max_contraction_velocity>`). Also, the workflows that involve manually editing XML files should hopefully be the same workflows that preserve XML comments (namely, I'm thinking of using Models/setup files in the distribution's Models folder, and the GUI's XML Browser). Surely, if one edits an XML file in MATLAB using the OpenSim API and calls `Object::print()`, they will still see XML comments. And we want to discourage using XML libraries directly to edit the files.

We can also ensure that the error message obtained when specifying a path to an object of the wrong type is very clear. For example, "The component at path '/foo/bar' is of type Ground, but must be of type Body." (I didn't check the code yet to see if such an error message is already in place).

chrisdembia replied:
> Previously, the Component had a property list of Connectors, which kept Connectors grouped.

We may be able to edit the serialization code to first serialize properties whose names contain `connectee_name`. What do you think?

I agree that this is definitely a downside, but if we must live with it: I feel that the proposed design has fewer downsides than the alternatives.

aseth1 replied:
>We can also ensure that the error message obtained when specifying a path to an object of the wrong type is very clear. For example, "The component at path '/foo/bar' is of type Ground, but must be of type Body." (I didn't check the code yet to see if such an error message is already in place).

I think that would be very helpful. One can turn to the *XMLBrowser* in the GUI or the *Doxygen* to learn of the type in the few instances that the comments are wiped out.  So, I am OK with this change.

aseth1 replied:
I have not thought it through completely, but what if Component maintained list properties for `Connector` and `Input` `connectee_name`s?

chrisdembia replied:
I have checked the error messages to display the required type in case that the connectee cannot be found.

chrisdembia commented:
> Might be easier to discuss in person. 

I have responded to your comments. I do still want to chat in person, but I decided to also type out my responses for record-keeping.

aseth1 commented:
It would also be good to get feedback from the group (@aymanhab @jimmyDunne  @jenhicks @tkuchida @carmichaelong) on the:

1. XML format of `Connector` (which will be renamed Socket in a future PR according to #1219) and `Input`
2. different order of `connectee_name` properties from their uniform listing at the top of the component in XML

before making any further changes.

chrisdembia commented:
> It would also be good to get feedback from the group

Good point. We could discuss in the dev meeting this Thursday?

chrisdembia commented:
Next steps:

- Move property construction out of the macro and into the `constructConnector()`, etc methods.
- Discuss with team (dev meeting?) about the ordering of properties in XML (should inputs be grouped together?).
- Improve error/documentation about connectee_names to quantities of the incorrect type.

chrisdembia commented:
Summary from dev meeting: Try to achieve the following format:

```xml
<MyComponent>
  <connectors>
  <!--Path to a Component to satisfy the Connector 'childFoo' of type Foo (description: The frame to which this geometry is attached. Note, being connected to a Frame means its transform is used to position this Geometry.).--> 
  <connector_frame_connectee_name>ground</connector_frame_connecteee_name>
  </connectors>
</MyComponent>
```

carmichaelong commented:
I think a further point is that if there is now a `<connectors>` section, then maybe we can further shorten the most inner tag without ambiguity? For instance, it could now be `<frame_connectee_name>` or maybe something shorter (but only if it's clear).

chrisdembia commented:
Ah yes good point. I remember that but I forgot to put that in the above comment. Thanks for the reminder.

chrisdembia commented:
@aseth1 this is ready for review again. I have:

1. put the construction of the connectee_name property within the `constructConnector()`, etc. methods,

2. renamed `AbstractConnector::finalizeFromProperties()` to `AbstractConnector::checkConnecteeNameProperty()`, and

3. added tests to make sure an exception is thrown for type mismatches (something exists at the connectee_name path, but it's not of the correct type).

aseth1 commented:
Would also be great to get feedback on impact of serialization and connection management via the API (@carmichaelong ?) and also from the GUI perspective (@aymanhab) .

aymanhab commented on OpenSim/Common/ComponentConnector.h:
```diff
@@ -295,38 +333,38 @@ class Connector : public AbstractConnector {
+
+        // check if the absolute pathname is just /name
+        if (objPathName.compare("/" + objT->getName()) == 0) { //exact match
+            // in which case we likely are connecting to an orphan
```
I thought we got rid of these explicit "/" and delegated this to PathName, is that possible here? Same with handling of "(",":" etc.

chrisdembia replied:
I have removed the explicit "/". @aseth1 it might be good for you to check this change, as I think you wrote this code originally.

aymanhab commented on OpenSim/Common/Object.h:
```diff
@@ -69,6 +69,10 @@ template class OSIMCOMMON_API OpenSim::ArrayPtrs<OpenSim::Object>;
+// Forward-declare SimTK types.
+namespace SimTK {
+    typedef Vec<2, Vec3> SpatialVec;
+}
```
Why were these added here?

chrisdembia replied:
Done.

aymanhab commented:
Finished with review. Minor suggestions/clarifications indicated above. An example that shows a minimal model with input serialization would be helpful. Thanks @chrisdembia Nicely done.

chrisdembia commented on Bindings/Java/Matlab/tests/testConnectorsInputsOutputs.m:
```diff
@@ -43,7 +43,7 @@
 
 % Access (and iterate through) a component's connectees as objects.
-expectedNames = {'ground', 'block'};
+expectedNames = {'block', 'ground'};
```
> If the latter are we guaranteed uniqueness?

Ah...there is an issue here. I will add a check to see if a connector/input with the same name already exists, and throw an exception if so.  Related: #859.

chrisdembia replied:
I have added thorough checks for uniqueness, and corresponding tests.

chrisdembia commented:
@aymanhab I have addressed your comments.

aymanhab commented on Bindings/Python/examples/wiring_inputs_and_outputs_with_TableReporter.py:
```diff
@@ -0,0 +1,99 @@
+# You can open the .sto file in a text editor and see that both outputs
+# (position of body's origin, and position of system mass center) are the same.
+
+
```
Very nice test @chrisdembia I would have created a new variable roundtrip_model to deserialize just for clarity. and also query the reporter upon deserialization that it has the right connections/inputs.  You went a step further to run a simulation which is great but is beyond the task at hand (of serializing/deserializing inputs. We can't verify correctness unless we run and check output file which I'm sure you did but may be a  bit harder to maintain. Thoughts?

chrisdembia replied:
Good points. I will rename the variable for clarity. I prefer `deserialized_model`. What do you think?

This file is an example, and so I don't really want it to look like a test (that is why I run a simulation). I had thought you were looking for a user-facing illustration of what this feature looks like. This new feature is thoroughly tested in C++ (testComponentInterface). If you'd like, I can also add a python (or Java) test.

aymanhab replied:
I like deserialized_model better as well. Go for it. 
As to what to do after deserialization I agree not to test/assert but you may want to modify something upon deserialization  that a user would possibly do ( and to show users how to recover access to the input). A Java example (for Matlab users) would be perfect, but if your environment isn't set up we can do later when other related open issues are resolved. Thanks.

chrisdembia replied:
Sounds good. Should I create a Java test or a MATLAB example? Or, a MATLAB test?

aymanhab commented:
Test failure appears unrelated. Finished with re-review, looking very good. I'm ok with merging as is or after addressing my last comment on the test, unless @aseth1 had other reservations.

aymanhab commented:
I d say matlab example

On Nov 21, 2016 10:40 AM, "Christopher Dembia" <notifications@github.com>
wrote:

> *@chrisdembia* commented on this pull request.
> ------------------------------
>
> In Bindings/Python/examples/wiring_inputs_and_outputs_
> with_TableReporter.py
> <https://github.com/opensim-org/opensim-core/pull/1340>:
>
> > +manager.setInitialTime(0)
> +manager.setFinalTime(1.0)
> +manager.integrate(state)
> +
> +# Now that the simulation is done, get the table from the TableReporter and
> +# write it to a file.
> +reporter = osim.TableReporterVec3.safeDownCast(model.getComponent('reporter'))
> +# This returns the TimeSeriesTableVec3 that holds the history of positions.
> +table = reporter.getTable()
> +# Create a FileAdapter, which handles writing to (and reading from) .sto files.
> +sto = osim.STOFileAdapterVec3()
> +sto.write(table, 'wiring_inputs_and_outputs_with_TableReporter.sto')
> +# You can open the .sto file in a text editor and see that both outputs
> +# (position of body's origin, and position of system mass center) are the same.
> +
> +
>
> Sounds good. Should I create a Java test or a MATLAB example? Or, a MATLAB
> test?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/opensim-org/opensim-core/pull/1340>, or mute the
> thread
> <https://github.com/notifications/unsubscribe-auth/ADyccMljhi8aEqMt9fKhnOQJc4DCJpcMks5rAeWUgaJpZM4KVaXi>
> .
>

chrisdembia commented:
@aymanhab I created a MATLAB example and improved the python example. I wanted to show editing of the alias in the examples, but ran into a bug wherein `setAlias()` does not affect the TableReporter's column labels. I have opened #1420 to track this bug. I do not want to tackle it here, because it will be easier to tackle in a follow-up PR that I have already started working on.

@aseth1 please review when you get the chance.

aseth1 commented on OpenSim/Common/ComponentConnector.h:
```diff
@@ -344,36 +381,99 @@ class Connector : public AbstractConnector {
+this input (that is, the connectees). The syntax for the connectee name
+property is as follows:
+@code
+<path/to/component/><output_name>[:<channel_name>][(<alias>)]
```
We discussed briefly,  should we separate the path from output name? As it stands we have to test what comes after the last '/' to see if it is an Output and not a Component.  That is to say paths always identify components so we have: `<path/to/component>|<output_name>[:<channel_name>][(<alias>)]`. 
We could leave for a followup PR unless you think it is easier to do now.

chrisdembia replied:
I started looking at implementing this, and I have decided that this change will fit much better in a subsequent PR.

aseth1 replied:
No problem. Thanks for looking into it.

chrisdembia replied:
So, I actually tried again and went through with the change. The component path and the output name are now separated by `|`. I decided it would be good to have this change in this PR, to avoid introducing an excess XML format update.

aseth1 replied:
Looks great to me.

aseth1 commented on OpenSim/Common/ComponentConnector.h:
```diff
@@ -344,36 +381,99 @@ class Connector : public AbstractConnector {
+protected:
+    /** Create a Connector that can only connect to Object of type T with 
+    specified name and stage at which it should be connected. Only Component
+    should ever construct this class.
```
should --> can?
As defined only a derived `Connector` or `Component` *can* invoke this constructor.

chrisdembia replied:
Done.

aseth1 commented on OpenSim/Common/XMLDocument.cpp:
```diff
@@ -427,6 +428,39 @@ void XMLDocument::addConnector(SimTK::Xml::Element& element,
+{
+    using ElementItr = SimTK::Xml::element_iterator;
+    
+    std::string typeName = "PhysicalFrame";
```
`typeName` is unused.

chrisdembia replied:
Removed.

aseth1 commented on OpenSim/Common/XMLDocument.cpp:
```diff
@@ -427,6 +428,39 @@ void XMLDocument::addConnector(SimTK::Xml::Element& element,
+    
+    // No longer want the old syntax for connectors.
+    componentElt.eraseNode(connectors_node);
+}
```
LGTM :+1:

aseth1 commented on OpenSim/Tests/Components/testComponents.cpp:
```diff
@@ -212,29 +227,66 @@ void testComponent(const Component& instanceToTest)
     }
-
+    
+    // Now make sure Inputs are satisfied.
```
Great addition @chrisdembia!!

chrisdembia commented on Applications/Analyze/test/subject01_adjusted_spring.osim:
```diff
@@ -1,1611 +1,1444 @@
 <?xml version="1.0" encoding="UTF-8" ?>
-<OpenSimDocument Version="30503">
+<OpenSimDocument Version="30000">
```
I wish that were the case. Here is part of the commit message for the commit in which I reverted the model:

> This commit reverts the model used in the testStaticOptimization passive force
test to version 30000. Previously, the model was at version 30503, but we
changed the Station connector's name from reference_frame to parent_frame after
30503 was first created.

This model file uses the outdated `reference_frame` as the name for Station's Connector. Unfortunately, there is no updateFromXML code that converts `reference_frame` to `parent_frame`. So it is not good to have this intermediate model version in the repo (we should also separately implement `Station::updateFromXMLNode()`).

chrisdembia replied:
I can alternatively rename `reference_frame` to `parent_frame` everywhere within the model file.

aseth1 replied:
Thanks for explanation. No need to change.

chrisdembia commented:
> I am glad to see the checks for name uniqueness, but I am still unclear as to why the order changed. Is it because child_frame connectee name appears before the parent_frame's in the property table (ordered alphabetically?) whereas the property list of connectors maintained order of elements as they were added?

This is because connectors are stored in an `std::map` (alphabetically). We discussed this during your initial review.

aseth1 commented on OpenSim/Auxiliary/auxiliaryTestFunctions.h:
```diff
@@ -127,6 +127,34 @@ do { \
+        throw OpenSim::Exception("TESTING: Expected exception " \
+            #EXPECTED_EXCEPTION " but no exception thrown."); \
+    } \
+} while(false) 
```
Makes sense. Thanks!

chrisdembia commented on Bindings/Java/Matlab/tests/testConnectorsInputsOutputs.m:
```diff
@@ -105,15 +105,15 @@
     numConnectees = rep.getInput(names.get(i)).getNumConnectees();
     assert(numConnectees == 4);
     for j = 0:(numConnectees - 1)
-        assert(strcmp(rep.getInput(names.get(i)).getAnnotation(j), ...
```
Updated to test label.

chrisdembia commented:
@aseth1 I have addressed your comments. Waiting for the tests to pass...

aseth1 commented:
Verified that CI failure is due to `testScale` that was just fixed. I will merging now.

chrisdembia commented:
Thank you.
